src/soter/openssl/soter_ec_key_3.c

```diff
@@ -0,0 +1,439 @@
+    }
+
+    if (!EVP_PKEY_get_octet_string_param(pkey,
+                                         param_name,
+                                         (unsigned char*)(key + 1),
+                                         output_length - sizeof(soter_container_hdr_t),
+                                         &serialized_len)) {
+        res = SOTER_INVALID_PARAMETER;
+        goto err;
+    }
+
+    if (serialized_len != output_length - sizeof(soter_container_hdr_t)) {
+        res = SOTER_INVALID_PARAMETER;
+        goto err;
+    }
+
+    memcpy(key->tag, ec_pub_key_tag(curve_str), SOTER_CONTAINER_TAG_LENGTH);
+    key->size = htobe32(output_length);
+    soter_update_container_checksum(key);
+    *key_length = output_length;
+    res = SOTER_SUCCESS;
+
+err:
+
+    return res;
+}
+
+soter_status_t soter_engine_specific_to_ec_priv_key(const soter_engine_specific_ec_key_t* engine_key,
+                                                    soter_container_hdr_t* key,
+                                                    size_t* key_length)
+{
+    EVP_PKEY* pkey = (EVP_PKEY*)engine_key;
+    const bool compressed = true;
+    soter_status_t res;
+    size_t output_length;
+    char curve_str[16];
+    BIGNUM* d = NULL;
+
+    if ((!key_length) || (EVP_PKEY_EC != EVP_PKEY_id(pkey))) {
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    // Curve identifier as string (SN_X9_62_prime256v1, SN_secp384r1, SN_secp521r1 etc)
+    if (!EVP_PKEY_get_utf8_string_param(pkey,
+                                        OSSL_PKEY_PARAM_GROUP_NAME /* "group" */,
+                                        curve_str,
+                                        sizeof(curve_str),
+                                        NULL)) {
+        res = SOTER_INVALID_PARAMETER;
+        goto err;
+    }
+
+    if (!is_curve_supported(curve_str)) {
+        res = SOTER_INVALID_PARAMETER;
+        goto err;
+    }
+
+    /*
+     * Note that we use a buffer suitable for a public key to store a private
+     * key. This was a historical mistake, now preserved for compatibility.
+     */
+    output_length = ec_pub_key_size(curve_str, compressed);
+    if ((!key) || (output_length > *key_length)) {
+        *key_length = output_length;
+        res = SOTER_BUFFER_TOO_SMALL;
+        goto err;
+    }
+
+    *key_length = output_length;
+
+    if (!EVP_PKEY_get_bn_param(pkey, OSSL_PKEY_PARAM_PRIV_KEY /* "priv" */, &d)) {
+        res = SOTER_INVALID_PARAMETER;
+        goto err;
+    }
+
+    if ((output_length - sizeof(soter_container_hdr_t))
+        != bn_encode(d, (unsigned char*)(key + 1), output_length - sizeof(soter_container_hdr_t))) {
+        res = SOTER_FAIL;
+        goto err;
+    }
+
+    memcpy(key->tag, ec_priv_key_tag(curve_str), SOTER_CONTAINER_TAG_LENGTH);
+    key->size = htobe32(output_length);
+    soter_update_container_checksum(key);
+    *key_length = output_length;
+    res = SOTER_SUCCESS;
+
+err:
+
+    return res;
```

Comment:
we should zeroize `d` if it is not null. For example if we exported key to the `d` but took error on `bn_encode` operation

Reply:
Done, using `BN_clear_free()` to bzero the value as well